tests/sys_execve/src/test_sys_execveat.c

```diff
@@ -0,0 +1,376 @@
+        0};
+    syscall(SYS_execveat, dirfd_bin, pathname, argv, 0, 0);
+}
+
+// Pathname is an empty string, flag set to AT_EMPTY_PATH
+// And dirfd points to an executable
+void test_execveat_empty_pathname_and_flag_at_empty_path_and_valid_dirfd_then_ok()
+{
+    printf("Running "
+           "test_execveat_empty_pathname_and_flag_at_empty_path_and_valid_"
+           "dirfd_then_ok\n");
+
+    // This program should run instead
+    const int dirfd_echo = open("/bin/echo", O_RDONLY);
+    const char* pathname = "";
+    char* const argv[] = {pathname,
+                          "test_execveat_empty_pathname_and_flag_at_empty_path_"
+                          "and_valid_dirfd_then_ok",
+                          0};
+    syscall(SYS_execveat, dirfd_echo, pathname, argv, 0, AT_EMPTY_PATH);
+}
+
+// Pathname is an empty string, flag NOT set to AT_EMPTY_PATH
+// And dirfd points to any executable
+// Same as invalid pathname
+void test_execveat_empty_pathname_and_empty_flag_and_any_dirfd_then_enoent()
+{
+    printf("Running "
+           "test_execveat_empty_pathname_and_empty_flag_and_any_dirfd_then_"
+           "enoent\n");
+
+    const int dirfd_bin = open("/bin/echo", O_RDONLY);
+    const char* pathname = "";
+    char* const argv[] = {pathname, "should not see this", 0};
+    int ret = syscall(SYS_execveat, dirfd_bin, pathname, argv, 0, 0);
+
+    expect(-1, ret, "execveat should return -1");
+    expect(ENOENT, errno, "errno should be ENOENT");
+    exit(0);
+}
+
+// Pathname is an empty string, flag set to AT_EMPTY_PATH
+// And dirfd points to a directory
+void test_execveat_empty_pathname_and_flag_at_empty_path_and_dirfd_directory_then_eacces()
+{
+    printf("Running "
+           "test_execveat_empty_pathname_and_flag_at_empty_path_and_dirfd_"
+           "directory_then_eacces\n");
+
+    const int dirfd_echo = open("/bin", O_RDONLY);
+    const char* pathname = "";
+    char* const argv[] = {pathname, "should not see this", 0};
+    int ret =
+        syscall(SYS_execveat, dirfd_echo, pathname, argv, 0, AT_EMPTY_PATH);
+
+    expect(-1, ret, "execveat should return -1");
+    expect(EACCES, errno, "errno should be EACCES");
+    exit(0);
+}
+
+// symlink as part of pathname
+void test_execveat_symlink_in_path_and_symlink_nofollow_flag_then_ok()
+{
+#ifdef TEST_SYMLINK
+    printf("Running "
+           "test_execveat_symlink_in_path_and_symlink_nofollow_flag_then_ok\n");
+
+    const int dirfd = AT_FDCWD;
+    const char* pathname = "/symlink/to_bin/echo";
+    char* const argv[] = {
+        pathname,
+        "test_execveat_symlink_in_path_and_symlink_nofollow_flag_then_ok",
+        0};
+
+    syscall(SYS_execveat, dirfd, pathname, argv, 0, AT_SYMLINK_NOFOLLOW);
+#else
+    printf("Skipping "
+           "test_execveat_symlink_in_path_and_symlink_nofollow_flag_then_ok\n");
+    exit(0);
+#endif
+}
+
+// last component of pathname is symlink
+void test_execveat_symlink_end_path_and_symlink_nofollow_flag_then_eloop()
```

Comment:
I have never seen a function name that is so descriptive. :-)

In practice, keep the name short yet descriptive. If you need to explain more what does it do, write a comment above the function declaration.

Reply:
Thanks. Will keep the current names as is, but will use better naming in the future :)